core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/BavetConstraintSessionFactory.java

```diff
@@ -42,19 +55,44 @@ public BavetConstraintSessionFactory(SolutionDescriptor<Solution_> solutionDescr
     public BavetConstraintSession<Solution_, Score_> buildSession(boolean constraintMatchEnabled,
             Solution_ workingSolution) {
         ScoreDefinition<Score_> scoreDefinition = solutionDescriptor.getScoreDefinition();
+        AbstractScoreInliner<Score_> scoreInliner = AbstractScoreInliner.buildScoreInliner(scoreDefinition,
+                constraintMatchEnabled);
+
         Score_ zeroScore = scoreDefinition.getZeroScore();
-        // Extract constraint weights, excluding constraints where weight is zero.
-        Map<BavetConstraint<Solution_>, Score_> constraintToWeightMap =
-                constraintList.stream()
-                        .map(constraint -> {
-                            Score_ weight = constraint.extractConstraintWeight(workingSolution); // Expensive, only do once.
-                            return new Object[] { constraint, weight };
-                        })
-                        .filter(constraintAndWeight -> !constraintAndWeight[1].equals(zeroScore))
-                        .collect(Collectors.toMap(
-                                constraintAndWeight -> (BavetConstraint<Solution_>) constraintAndWeight[0],
-                                constraintAndWeight -> (Score_) constraintAndWeight[1]));
-        return new BavetConstraintSession<>(constraintMatchEnabled, scoreDefinition, constraintToWeightMap);
+        Set<BavetAbstractConstraintStream<Solution_>> constraintStreamSet = new LinkedHashSet<>(constraintList.size() * 10);
+        Map<Constraint, Score_> constraintWeightMap = new HashMap<>(constraintList.size());
+        for (BavetConstraint<Solution_> constraint : constraintList) {
+            // Expensive, only do once.
+            Score_ constraintWeight = constraint.extractConstraintWeight(workingSolution);
+            // Node sharing happens earlier, in BavetConstraintFactory#share(Stream_).
+            // Filter out nodes that only lead to constraints with zero weight
+            if (!constraintWeight.equals(zeroScore)) {
+                // Relies on BavetConstraintFactory#share(Stream_) occurring for all constraint stream instances
+                // to ensure there are no 2 equal constraint streams (with different child stream lists).
+                constraint.collectActiveConstraintStreams(constraintStreamSet);
+                constraintWeightMap.put(constraint, constraintWeight);
+            }
+        }
+        NodeBuildHelper<Score_> buildHelper = new NodeBuildHelper<>(constraintStreamSet, constraintWeightMap, scoreInliner);
+        // Reverse constraintStreamSet order to create downstream nodes first
+        List<BavetAbstractConstraintStream<Solution_>> reversedConstraintStreamList = new ArrayList<>(constraintStreamSet);
+        Collections.reverse(reversedConstraintStreamList);
+        for (BavetAbstractConstraintStream<Solution_> constraintStream : reversedConstraintStreamList) {
+            constraintStream.buildNode(buildHelper);
+        }
```

Comment:
You gave a very nice explanation for why this is here, when we had our chat.
Maybe that explanation should also be here, as a comment?